module_3/task_2/hw2.py

```diff
@@ -0,0 +1,27 @@
+"""
+Write a function that accepts another function as an argument. Then it
+should return such a function, so the every call to initial one
+should be cached.
+def func(a, b):
+    return (a ** b) ** 2
+cache_func = cache(func)
+some = 100, 200
+val_1 = cache_func(*some)
+val_2 = cache_func(*some)
+assert val_1 is val_2
+"""
+from collections.abc import Callable
+
+
+def cache(func: Callable) -> Callable:
+    cached_results = {}
+
+    def cache_func(*args):
+        if args in cached_results:
+            return cached_results
+        else:
+            for arg in args:
+                cached_results[arg] = func(*args)
+        return cached_results
+
+    return cache_func
```

Comment:
Okay, test passes, but that's my fault - the test should be rewritten, because your solution doesn't really cache anything.
First - on line 20 you are checking if the `*args` (remember, this is a tuple) are key in cached_results dict. If they are - you are returning whole cache. But the cache really looks something like this:
```py
{(100, 200): 333, (200, 300): 555, ...}
```
So you should return some specific entry from it.
And on the line 24 you shouldn't really iterate over args. You should cache result of a function call with arguments `*args`

Reply:
btw, you can add this to the end of the test, it should pass with a correct solution
```py
val_3 = cache_func(200, 300)
assert val_1 == (some[0] ** some[1]) ** 2
assert val_2 == (some[0] ** some[1]) ** 2
assert val_2 is not val_3
```

Reply:
Fixed, test passed with additional assertions:
![image](https://github.com/sergei820/PythonTAProgramEpam/assets/56273100/8a065a47-5922-4b8d-91ae-8cfc4278dfd6)